yt/data_objects/construction_data_containers.py

```diff
@@ -764,26 +763,54 @@ def _fill_sph_particles(self, fields):
                     dens = chunk[(ptype,'density')].in_base("code").d
                     field_quantity = chunk[field].d
 
-                    pixelize_sph_kernel_arbitrary_grid(dest, px, py, pz, hsml,
-                                    mass, dens, field_quantity, bounds, pbar)
+                    pixelize_sph_kernel_arbitrary_grid(buff, px, py, pz, hsml,
+                                    mass, dens, field_quantity, bounds,
+                                    pbar=pbar)
                     if normalize:
-                        pixelize_sph_kernel_arbitrary_grid(dest_den, px, py, pz,
-                                    hsml, mass, dens, np.ones(hsml.shape[0]),
-                                    bounds)
+                        pixelize_sph_kernel_arbitrary_grid(buff_den, px, py, pz,
+                                    hsml, mass, dens, np.ones(dens.shape[0]),
+                                    bounds, pbar=pbar)
 
                 if normalize:
-                    normalization_3d_utility(dest, dest_den)
+                    normalization_3d_utility(buff, buff_den)
 
                 fi = self.ds._get_field_info(field)
-                self[field] = self.ds.arr(dest, fi.units)
+                self[field] = self.ds.arr(buff, fi.units)
                 pbar.close()
 
         if(smoothing_style == "gather"):
+            num_neighbors = getattr(self.ds, 'num_neighbors', 32)
             for field in fields:
-                buff = np.zeros(self.ActiveDimensions, dtype="float64")
-
-                pixelize_sph_gather(buff, bounds, self.ds, field,
-                                    ptype, normalize=normalize)
+                buff = np.zeros(size, dtype="float64")
+
+                tree_positions = []
+                hsml = []
+                pmass = []
+                pdens = []
+                quantity_to_smooth = []
+                for chunk in self.ds.all_data().chunks([field],"io"):
+                    tree_positions.append(chunk[(ptype,
+                                                 'particle_position')].in_base("code").d)
+                    hsml.append(chunk[(ptype,
+                                       'smoothing_length')].in_base("code").d)
+                    pmass.append(chunk[(ptype,
+                                       'particle_mass')].in_base("code").d)
+                    pdens.append(chunk[(ptype,
+                                       'density')].in_base("code").d)
+                    quantity_to_smooth.append(chunk[field].d)
+
+                kdtree = self.ds.index.kdtree
+                tree_positions = np.concatenate(tree_positions)[kdtree.idx, :]
+                hsml = np.concatenate(hsml)[kdtree.idx]
+                pmass = np.concatenate(pmass)[kdtree.idx]
+                pdens = np.concatenate(pdens)[kdtree.idx]
+                quantity_to_smooth = np.concatenate(quantity_to_smooth)[kdtree.idx]
```

Comment:
I think it might make sense to refactor the boilerplate above and in `cartesian_coordinates.py` into a single function that takes a dataset and returns the concatenated arrays.

Reply:
yep, we do use it in a few places